src/clj/puppetlabs/puppetserver/certificate_authority.clj

```diff
@@ -475,6 +494,51 @@
    (utils/subject-key-identifier
      ca-public-key false)])
 
+(schema/defn read-compile-masters
+    "Returns a list of compile masters from the specified file.
+     Note that this method is agnostic of the file contents which could either
+     be node names (FQDNs) or certificate serial numbers so long as they are
+     organized as one item per line.
+     Likewise this method could/will be re-used"
+    [cm-file :- schema/Str]
+    (line-seq (io/reader cm-file)))
+
+(schema/defn generate-compile-master-serials
+  "Given a list of compile master nodes it will create a file containing 
+   serial numbers of their certificates (listed on separate lines).
+   Compile masters is expected have at least one entry (MoM)"
+   [cm-nodes :- [schema/Str]
+    compile-masters-serials-file :- schema/Str
+    signeddir :- schema/Str]
+
+   ;; TODO: We may want to account for the possibility of compile masters to
+   ;; be empty (by mistake, accident). In which case this can use the name
+   ;; of the current node as the MoM as a fallback
```

Comment:
Is this feature something that will be opted into (I assume not)?

Reply:
Am thinking not. Though we could make it configurable if there might be issues with cloud deployments using ephemeral compile masters and there are concerns around infra-crl not accurately reflecting revocation status of all compile masters. But then they will have to deal with potentially large CRLs (since even agents would be ephemeral and thus large number of revocations). 

As I think about it, I feel it should not be too difficult to stamp an extension indicating the role of the node. Though need to think if it can also be done reliably in a pure foss deployment (with no classifier etc.)

Reply:
So, I think the custom extension denoting the puppet infra role would be valuable and itself not a lot of work, but it would require coordination across our products and making a policy about how to create certs for puppet infra, which we don't have currently and have always wavered on. That level of decisions and coordinating probably make it a significantly harder thing to implement then this approach.

Reply:
Agree. It occurs to me the scheme I was thinking (of automating the extension stamping transparently) actually might not work. It breaks down in case a node get repurposed for some other role but continues to use the original certificate. But if you were alluding to some more explicit manner of how infra certificates are created  (that can avoid above situations) then that should work sans the coordination challenges.